rfcs/text/0008_service_status.md

```diff
@@ -0,0 +1,251 @@
+- Start Date: 2020-03-07
+- RFC PR: (leave this empty)
+- Kibana Issue: (leave this empty)
+
+# Summary
+
+A set API for describing the current status of a system (Core service or plugin)
+in Kibana.
+
+# Basic example
+
+```ts
+core.status.set({
+  level: ServiceStatusLevel.degraded,
+  summary: `APM indices cannot be created`,
+});
+```
+
+# Motivation
+
+Kibana should do as much possible to help users keep their installation in a working state. This includes providing as much detail about components that are not working as well as ensuring that failures in one part of the application do not block using other portions of the application.
+
+In order to provide the user with as much detail as possible about any systems that are not working correctly, the status mechanism should provide excellent defaults in terms of expressing relationships between services and presenting detailed information to the user.
+
+# Detailed design
+
+## Types
+
+```ts
+/**
+ * The current status of a service at a point in time.
+ */
+interface ServiceStatus {
+  /** The current availability level of the service. */
+  level: ServiceStatusLevel;
+  /** A high-level summary of the service status. */
+  summary?: string;
+  /** A more detailed description of the service status. */
+  detail?: string;
+  /** A URL to open in a new tab about how to resolve or troubleshoot the problem */
+  documentationUrl?: string;
+  /** Any JSON-serializable data to be included in the HTTP API response. */
+  meta?: object;
+}
+
+/**
+ * The current "level" of availability of a service.
+ */
+enum ServiceStatusLevel {
+  available,   // everything is working!
+  degraded,    // some features may not be working
+  unavailable, // the service is unavailable, but other functions should work
+  fatal        // block all user functions and display the status page
```

Comment:
Shouldn't `fatal` be core (+inheritance) reserved? Does it make sense for a plugin to manually `set` it's state to `fatal`?

Reply:
There's at least one use case I can think of where a plugin may want to force Kibana into a `fatal` mode: Audit Logging. There is a soft requirement that audit logging have a config option that makes Kibana unusable if logs cannot be written to disk.

@jportner would you prefer this be implemented by actually crashing the server process or by blocking user features?

It's also possible the security plugin could want to use this, but the security plugin's hooks elsewhere may be sufficient.

If neither of these use cases are valid, I agree we should limit this level to Core only.

Reply:
I'm going to go ahead and make this Core-only. If we find later that a plugin truly needs it, we can adjust.